### PR TITLE
Minor: Upgrade docs for `PhysicalExpr::{propagate_constraints, evaluate_bounds}`

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

N/A

## Rationale for this change
While reviewing https://github.com/apache/arrow-datafusion/pull/7804#pullrequestreview-1674090033 I spent time figuring out in more detail what these functions did. I figured I would save myself (and hopefully other future readers) some time in the future

## What changes are included in this PR?

Add doc comments and examples for `PhysicalExpr::{propagate_constraints, evaluate_bounds}`

## Are these changes tested?
Docs

### CODE REVIEW

File: datafusion/physical-expr/src/physical_expr.rs
Comment (alamb) [inline datafusion/physical-expr/src/physical_expr.rs:115]: @metesynnada  and @berkaysynnada -- I would appreciate it if you could double check this example. I think this is the kind of thing `propagate_constraints` can do, but I am not 100% sure I got the example correct

File: datafusion/physical-expr/src/physical_expr.rs
Comment (berkaysynnada) [inline datafusion/physical-expr/src/physical_expr.rs:119]:
```suggestion
    /// would return `[a: [0, 2], b: [2, 4]]` as `b` must be at least 2 to
```

File: datafusion/physical-expr/src/physical_expr.rs
Comment (berkaysynnada) [reply datafusion/physical-expr/src/physical_expr.rs:115]: The example you gave is clear and correct, thank you ☺️ 
For more detail:
- For plus operation, specifically, we would first do
- [a_lower, a_upper] <- ([expression_lower, expression_upper] - [b_lower, b_upper]) ∩ [a_lower, a_upper], and then
- [b_lower, b_upper] <- ([expression_lower, expression_upper] -  [a_lower, a_upper]) ∩ [b_lower, b_upper].

**Comment by alamb** [comment]
@metesynnada  or @ozankabak  could I trouble you for a review of this PR ? I can't merge it until a committer has marked it as approved

**Review by ozankabak — APPROVED** [review]
🚀

**Comment by alamb** [comment]
Thank you @ozankabak

**Pull request merged**
